sql/query/expr.go

```diff
@@ -0,0 +1,660 @@
+	if err != nil {
+		return null, err
+	}
+	if v, err := right.Bool(); err != nil {
+		return null, err
+	} else if v {
+		return boolToDatum(true), nil
+	}
+	return boolToDatum(false), nil
+}
+
+func evalNotExpr(expr *parser.NotExpr, env Env) (sqlwire.Datum, error) {
+	d, err := EvalExpr(expr.Expr, env)
+	if err != nil {
+		return null, err
+	}
+	v, err := d.Bool()
+	if err != nil {
+		return null, err
+	}
+	return boolToDatum(!v), nil
+}
+
+func evalRangeCond(expr *parser.RangeCond, env Env) (sqlwire.Datum, error) {
+	// TODO(pmattis): This could be more efficient or done ahead of time.
+	d, err := EvalExpr(&parser.AndExpr{
+		Left: &parser.ComparisonExpr{
+			Operator: parser.GE,
+			Left:     expr.Left,
+			Right:    expr.From,
+		},
+		Right: &parser.ComparisonExpr{
+			Operator: parser.LE,
+			Left:     expr.Left,
+			Right:    expr.To,
+		},
+	}, nil)
+	if err != nil {
+		return null, err
+	}
+	if expr.Not {
+		*d.BoolVal = !*d.BoolVal
+	}
+	return d, nil
+}
+
+func evalNullCheck(expr *parser.NullCheck, env Env) (sqlwire.Datum, error) {
+	d, err := EvalExpr(expr.Expr, env)
+	if err != nil {
+		return null, err
+	}
+	v := d.IsNull()
+	if expr.Not {
+		v = !v
+	}
+	return boolToDatum(v), nil
+}
+
+// Prepare the arguments for a comparison operation. The returned arguments
+// will have the same type.
+func prepareComparisonArgs(left, right sqlwire.Datum) (opType, sqlwire.Datum, sqlwire.Datum, error) {
+	// If both arguments are strings (or string-like), compare as strings.
+	if (left.BytesVal != nil || left.StringVal != nil) &&
+		(right.BytesVal != nil || right.StringVal != nil) {
+		return stringOp, left.ToString(), right.ToString(), nil
+	}
+
+	// If both arguments are uints, compare as unsigned.
+	if left.UintVal != nil && right.UintVal != nil {
+		return uintOp, left, right, nil
+	}
+
+	var err error
+
+	// If both arguments are integers (signed or unsigned), compare as integers.
+	if (left.BoolVal != nil || left.IntVal != nil || left.UintVal != nil) &&
+		(right.BoolVal != nil || right.IntVal != nil || right.UintVal != nil) {
+		left, err = left.ToInt()
+		if err != nil {
+			return intOp, null, null, err
+		}
+		right, err = right.ToInt()
+		if err != nil {
+			return intOp, null, null, err
+		}
+		return intOp, left, right, nil
+	}
+
+	// In all other cases, compare as floats.
+	left, err = left.ToFloat()
+	if err != nil {
+		return intOp, null, null, err
+	}
+	right, err = right.ToFloat()
+	if err != nil {
+		return intOp, null, null, err
+	}
+	return floatOp, left, right, nil
+}
+
+func evalComparisonExpr(expr *parser.ComparisonExpr, env Env) (sqlwire.Datum, error) {
+	left, err := EvalExpr(expr.Left, env)
+	if err != nil {
+		return null, err
+	}
+	right, err := EvalExpr(expr.Right, env)
+	if err != nil {
+		return null, err
+	}
+
+	op := expr.Operator
+	if op == parser.NullSafeEqual {
+		if left.IsNull() && right.IsNull() {
+			return boolToDatum(true), nil
+		}
+		op = parser.EQ
+	}
+
+	if left.IsNull() || right.IsNull() {
```

Comment:
This is part of the SQL standard. To compare with `NULL` you are supposed to to `<val> IS NULL` (which I haven't implemented yet). I think the idea is that `NULL` is a bit like `NaN`.